**Introdução**

In [2]:
print('Hello, World!')

Hello, World!


#### Devemos rodar nossos projetos de Machine Learning em GPU quando formos usar no Google Colab.

#### Kagle.com é onde devemos pegar nossos datasets. O que usaremos nessa aula é o "Diamonds".

**Importação de bibliotecas**

In [39]:
import pandas as pd ## estamos apelidando a biblioteca pandas para "pd"
from sklearn.model_selection import train_test_split

ModuleNotFoundError: No module named 'sklearn'

**Importação dos dados e criação do Dataframe**

In [4]:
df = pd.read_csv('diamonds.csv') ## estamos pegando o arquivo "diamonds.csv" para usar

In [5]:
type(df) ## qual o tipo de dado do nosso df?

pandas.core.frame.DataFrame

**Análise Exploratória**

In [6]:
# Linhas ---> Registros
# Colunas ---> Atributos

In [7]:
# Visualização dos primeiros registros
df.head(10)

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
0           1   0.23      Ideal     E     SI2   61.5   55.0    326  3.95   
1           2   0.21    Premium     E     SI1   59.8   61.0    326  3.89   
2           3   0.23       Good     E     VS1   56.9   65.0    327  4.05   
3           4   0.29    Premium     I     VS2   62.4   58.0    334  4.20   
4           5   0.31       Good     J     SI2   63.3   58.0    335  4.34   
5           6   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94   
6           7   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95   
7           8   0.26  Very Good     H     SI1   61.9   55.0    337  4.07   
8           9   0.22       Fair     E     VS2   65.1   61.0    337  3.87   
9          10   0.23  Very Good     H     VS1   59.4   61.0    338  4.00   

      y     z  
0  3.98  2.43  
1  3.84  2.31  
2  4.07  2.31  
3  4.23  2.63  
4  4.35  2.75  
5  3.96  2.48  
6  3.98  2.47  
7  4.11  2.53  
8  3.78  2.49  
9  4.05  2.39

In [8]:
# Dimensões do dataframe
df.shape

(53940, 11)

In [9]:
print(f'O dataframe tem {df.shape[0]} registros e {df.shape[1]} atributos.')

O dataframe tem 53940 registros e 11 atributos.


In [10]:
# Informações dos atributos do dataframe
# object ---> dados categóricos ---> valores em formato str (String)
# devemos converter as colunas para dados numéricos (label_encoder)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  53940 non-null  int64  
 1   carat       53940 non-null  float64
 2   cut         53940 non-null  object 
 3   color       53940 non-null  object 
 4   clarity     53940 non-null  object 
 5   depth       53940 non-null  float64
 6   table       53940 non-null  float64
 7   price       53940 non-null  int64  
 8   x           53940 non-null  float64
 9   y           53940 non-null  float64
 10  z           53940 non-null  float64
dtypes: float64(6), int64(2), object(3)
memory usage: 4.5+ MB


**É de extrema importancia que você separe os atributos (tabelas), filtre, para usar somente a que você quer e precisa usar.**

In [11]:
# Estatísticas descritivas dos atributos numéricos
df.describe() ## descrição dos valores

Unnamed: 0         carat         depth         table         price  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean   26970.500000      0.797940     61.749405     57.457184   3932.799722   
std    15571.281097      0.474011      1.432621      2.234491   3989.439738   
min        1.000000      0.200000     43.000000     43.000000    326.000000   
25%    13485.750000      0.400000     61.000000     56.000000    950.000000   
50%    26970.500000      0.700000     61.800000     57.000000   2401.000000   
75%    40455.250000      1.040000     62.500000     59.000000   5324.250000   
max    53940.000000      5.010000     79.000000     95.000000  18823.000000   

                  x             y             z  
count  53940.000000  53940.000000  53940.000000  
mean       5.731157      5.734526      3.538734  
std        1.121761      1.142135      0.705699  
min        0.000000      0.000000      0.000000  
25%        4.710000      4.720000      2.910000  
50%        5.700000      5.710000      3.530000  
75%        6.540000      6.540000      4.040000  
max       10.740000     58.900000     31.800000

Média, mediana e desvio padrão definem sua base de dados.

In [12]:
# (futuro) Estatísticas descritivas dos atributos categóricos

In [13]:
# Média aritmética da coluna de preços (price)
media_aritmetica = df['price'].mean()
print(f'A media aritmética dos preços é: {media_aritmetica: .2f}')

A media aritmética dos preços é:  3932.80


In [14]:
# Mediana da coluna de preços (price)
mediana = df['price'].median()
print(f'A mediana dos preços é: {mediana: .2f}')

A mediana dos preços é:  2401.00


In [15]:
# Desvio Padrão da coluna de preços (price)
desvio_padrao = df['price'].std()
print(f'O desvio padrão dos preços é: {desvio_padrao: .2f}')

O desvio padrão dos preços é:  3989.44


**Geração de amostras**

Dados da população:
Média Aritmética: US$ 3932.80
Mediana: US$ 2401.00
Desvio Padrão: US$ 3989.

In [ ]:
# Primeira amostra (5% dos valores iniciais ---> 2700 registros)
# Média aritmética, mediana e desvio padrão da coluna de preços (price)
a1 = df['price'].iloc[0:2700]

a1_media_aritmetica = a1.mean()
print(f'A media aritmética dos preços é: {a1_media_aritmetica: .2f}')

a1_mediana = a1.median()
print(f'A mediana dos preços é: {a1_mediana: .2f}')

a1_desvio_padrao = a1.std()
print(f'O desvio padrão dos preços é: {a1_desvio_padrao: .2f}')

A media aritmética dos preços é:  2709.73
A media aritmética dos preços é:  2940.00
A media aritmética dos preços é:  785.84


In [ ]:
# Segunda amostra (5% dos valores finais ---> 2700 registros)
# Média aritmética, mediana e desvio padrão da coluna de preços (price)
a2 = df['price'].iloc[51239:53939]

a2_media_aritmetica = a2.mean()
print(f'A media aritmética dos preços é: {a2_media_aritmetica: .2f}')

a2_mediana = a2.median()
print(f'A mediana dos preços é: {a2_mediana: .2f}')

a2_desvio_padrao = a2.std()
print(f'O desvio padrão dos preços é: {a2_desvio_padrao: .2f}')

A media aritmética dos preços é:  2365.68
A mediana dos preços é:  2517.00
O desvio padrão dos preços é:  578.95


In [ ]:
# Terceira amostra (5% dos valores centrais ---> 2700 registros)
# Média aritmética, mediana e desvio padrão da coluna de preços (price)
a3 = df['price'].iloc[25650:28350]

a3_media_aritmetica = a3.mean()
print(f'A media aritmética dos preços é: {a3_media_aritmetica: .2f}')

a3_mediana = a3.median()
print(f'A mediana dos preços é: {a3_mediana: .2f}')

a3_desvio_padrao = a3.std()
print(f'O desvio padrão dos preços é: {a3_desvio_padrao: .2f}')

A media aritmética dos preços é:  11601.85
A mediana dos preços é:  15618.00
O desvio padrão dos preços é:  7449.37


In [ ]:
# Amostra aleatória (5% dos valores ---> 2700 registros)
# Média aritmética, mediana e desvio padrão da coluna de preços (price)
a4 = df['price'].sample(frac=0.05, random_state=10) ## seed, valor que fixa a amostra

a4_media_aritmetica = a4.mean()
print(f'A media aritmética dos preços é: {a4_media_aritmetica: .2f}')

a4_mediana = a4.median()
print(f'A mediana dos preços é: {a4_mediana: .2f}')

a4_desvio_padrao = a4.std()
print(f'O desvio padrão dos preços é: {a4_desvio_padrao: .2f}')

A media aritmética dos preços é:  3834.44
A mediana dos preços é:  2401.00
O desvio padrão dos preços é:  3855.03


**Tarefa:** 

### Implementar um modelo preditivo de classificação para prever a categoria do corte (coluna 'cut') do diamante com base nas dimensões (colunas'x', 'y' e 'z')

In [ ]:
# Separação de dados de entrada X (features, variáveis independentes) 
# e dados de saída y (target, variável dependente)

In [31]:
# X = df[['x', 'y', 'z']]
X = df.iloc[:, -3:] # : ---> todos os registros, -3: ---> da antepenúltima até o final
X

x     y     z
0      3.95  3.98  2.43
1      3.89  3.84  2.31
2      4.05  4.07  2.31
3      4.20  4.23  2.63
4      4.34  4.35  2.75
...     ...   ...   ...
53935  5.75  5.76  3.50
53936  5.69  5.75  3.61
53937  5.66  5.68  3.56
53938  6.15  6.12  3.74
53939  5.83  5.87  3.64

[53940 rows x 3 columns]

In [32]:
y = df['cut']
y

0            Ideal
1          Premium
2             Good
3          Premium
4             Good
           ...    
53935        Ideal
53936         Good
53937    Very Good
53938      Premium
53939        Ideal
Name: cut, Length: 53940, dtype: object

In [37]:
# Separação de dados de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=11)

NameError: name 'train_test_split' is not defined

In [ ]:
X_train.shape

In [ ]:
X_test.shape

In [ ]:
y_test.shape

In [ ]:
y_train.shape

### Escolha o seu modelo/algoritmo de classificação

In [ ]:
# KNN
from sklearn.neighbors import KNeighborsClassifier

# Árvore de decisão 
from sklearn.tree import DecisionTreeClassifier

# Random Forest
from sklearn.ensemble import RandomForestClassifier

# Regressão Logística
from sklearn.linear_model import LogisticRegression

In [ ]:
# Modelo KNN
modelo_knn = KNeighborsClassifier(n_neighbors=5)

In [ ]:
# Treinamento do modelo
modelo_knn.fit(X_train, y_train)

In [ ]:
# Previsões do modelo KNN
y_predict = modelo_knn.predict(X_test)
y_predict

In [ ]:
# Avaliando modelo com base na acurácia
# y_test.values
from sklearn.metrics import accuracy_score

acuracia = accuracy_score(y_test, y_predict)
print(f'A acurácia do modelo KNN: {acuracia * 100: .2f}%')